File: node/iris_auto_sign_in.js
const crypto = require("crypto");
const { URL, URLSearchParams } = require("url");

class IrisAutoSignIn {
  constructor(secretKey, isProduction = false) {
    this.secretKey = secretKey;
    this.rootUrl = isProduction
      ? "https://iris.opynhealth.com"
      : "https://iris-test.opynhealth.com";
  }

  embedUrl(clientId, groupId, memberId, expiresAt) {
    return this._urlBuild(`${this.rootUrl}/embed`, clientId, groupId, memberId, expiresAt);
  }

  autoSignInUrl(clientId, groupId, memberId, expiresAt) {
    return this._urlBuild(`${this.rootUrl}/autoSignIn`, clientId, groupId, memberId, expiresAt);
  }

  _urlBuild(endpoint, clientId, groupId, memberId, expiresAt) {
    const expiresAtText = expiresAt.toISOString();
    const message = `${clientId}:${groupId}:${memberId}:${expiresAtText}`;

    // Create HMAC-SHA256 signature
    const hmac = crypto.createHmac("sha256", this.secretKey);
    hmac.update(message);
    const signature = hmac.digest("base64");

    // Construct the full URL with parameters
    const params = new URLSearchParams({
      clientId,
      groupId,
      memberId,
      expires: expiresAtText,
      signature,
    });

    return `${endpoint}?${params.toString()}`;
  }
}

// Example usage
const secretKey = "SECRET_KEY";
const clientId = "CLIENT_ID";
const groupId = "GROUP_ID";
const memberId = "MEMBER_ID";
const isProduction = false;
const expiresAt = new Date(Date.now() + 24 * 60 * 60 * 1000); // 24 hours from now

const asi = new IrisAutoSignIn(secretKey, isProduction);

console.log("Auto sign-in URL:");
console.log(asi.autoSignInUrl(clientId, groupId, memberId, expiresAt));
console.log("");

console.log("Embedded URL:");
console.log(asi.embedUrl(clientId, groupId, memberId, expiresAt));

File: python/iris_auto_sign_in.py

import hmac
import hashlib
import base64
from datetime import datetime, timedelta
from urllib.parse import urlencode

class IrisAutoSignIn:
    def __init__(self, secret_key, is_production=False):
        self.secret_key = secret_key
        self.root_url = "https://iris.opynhealth.com" if is_production else "https://test-iris.opynhealth.com"

    def embed_url(self, client_id, group_id, member_id, expires_at):
        return self._url_build(f"{self.root_url}/embed", client_id, group_id, member_id, expires_at)

    def auto_sign_in_url(self, client_id, group_id, member_id, expires_at):
        return self._url_build(f"{self.root_url}/autoSignIn", client_id, group_id, member_id, expires_at)

    def _url_build(self, endpoint, client_id, group_id, member_id, expires_at):
        expires_at_text = expires_at.isoformat()
        message = f"{client_id}:{group_id}:{member_id}:{expires_at_text}"
        
        # Create HMAC-SHA256 signature
        hmac_digest = hmac.new(self.secret_key.encode(), message.encode(), hashlib.sha256).digest()
        signature = base64.b64encode(hmac_digest).decode()

        # Construct the full URL with parameters
        params = {
            "clientId": client_id,
            "groupId": group_id,
            "memberId": member_id,
            "expires": expires_at_text,
            "signature": signature
        }

        return f"{endpoint}?{urlencode(params)}"

if __name__ == "__main__":
    secret_key = "SECRET_KEY"
    client_id = "CLIENT_ID"
    group_id = "GROUP_ID"  # Group ID or group code
    member_id = "MEMBER_ID"
    is_production = False
    expires_at = datetime.utcnow() + timedelta(hours=24)

    asi = IrisAutoSignIn(secret_key, is_production)

    print("## Python Example ##")
    print("Auto sign-in URL")
    auto_sign_in_url = asi.auto_sign_in_url(client_id, group_id, member_id, expires_at)
    print(auto_sign_in_url)
    print("")

    print("Embedded URL:")
    embed_url = asi.embed_url(client_id, group_id, member_id, expires_at)
    print(embed_url)
